src/main/scala/za/co/absa/commons/s3/S3Location.scala

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.commons.s3
+
+import scala.util.matching.Regex
+
+object S3Location {
+
+  /**
+   * Generally usable regex for validating S3 path, e.g. `s3://my-cool-bucket1/path/to/file/on/s3.txt`
+   * Protocols `s3`, `s3n`, and `s3a` are allowed.
+   * Bucket naming rules defined at [[https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html#bucketnamingrules]] are instilled.
+   */
+  private val S3LocationRx: Regex = "(s3[an]?)://([-a-z0-9.]{3,63})/(.*)".r
+
+  implicit class StringS3LocationExt(val path: String) extends AnyVal {
+
+    def toS3Location: Option[SimpleS3Location] = PartialFunction.condOpt(path) {
+      case S3LocationRx(protocol, bucketName, relativePath) => 
+        SimpleS3Location(protocol, bucketName, relativePath)
+    }
+
+    def isValidS3Path: Boolean = S3LocationRx.pattern.matcher(path).matches
+  }
+}
+
+trait S3Location {
```

Comment:
Why is this trait needed? What other implementations could exists except for `SimpleS3Location"? Which of these definitions are polymorphic?

Reply:
I tend to agree with Alex. I think a `case class S3Location` should be enough.

Reply:
There are cases, where other S3Locations are useful (e.g. in Atum, when working with S3 using the SDK, it is advantageous to have S3Location with Region. And there, the trait is useful as well, because you can pass any kind of S3Location (Simple or not) to have it enriched with Region. That also why I would not want to seal the trait.
I will think about it.

https://github.com/AbsaOSS/atum/blob/a5ee735a42c939ebf6f966ecad91f726e8a807af/atum-s3-sdk-extension/src/main/scala/za/co/absa/atum/persistence/s3/Regional.scala


> In that case the method `s3String` should also be polymorphic ...
Yes, I can see that now, I agree and will adjust it. Thank you.

Reply:
I understand what you are talking about, but I am trying to challenge the way the goal is (if) achieved.
```scala
case class FooLocaltion(foo: Foo, ....) extends S3Location
import S3LocationRegionImplicits._

val fooLoc = FooLocation(foo, ...)
val fooLocWithRegion = fooLoc.withRegion(regoin) // oops! it's not FooLocation any more!

// another imaginary example

import S3location._

val aSimpleLoc = "s3://blah/blah".toS3Location
val aProtectedLoc =  "protected://blah".toS3Location
val aRegionalLoc = "region:aware://....".toS3Location

aRegionalLoc.region // doesn't compile
aProtectedLoc.token // doesn't compile

```

Reply:
What if there is only one case class that represents the AWS spec and have all components in it. Then you can use traits to  mix in operations that are specific for certain "subtypes" of location, if you wish.

Reply:
I have really tried both (trait `S3Location` & `case class SimpleS3Location` vs just `S3Location`) and having another S3Location type by composition instead of _is-a_, is quite annoying (e.g. in Atum).
So, if it is ok with the trait, I would very much like to keep it - to be open for an extension as the open/closed principle suggests.

Reply:
None of these is expected
> `"protected://blah".toS3Location`
> `"region:aware://....".toS3Location`

Only `"s3://blah/blah".toS3Location` is needed. So, the {from|to} {String|SimpleS3Location} is the only parsing and serialization needed.

I have updated the namings to convey the meaning better, the parsing is now not done on the `S3Location` object (it obviously communicated that different kinds of locations could have been parsed - that is not true, so now only the `SimpleS3Location` can parse it. So its `"s3://bla/bla".toSimpleS3Location` and it directly returns the SimpleS3Location (option of it or directly when used using `apply`). So here, we have the most simple case class usage as probably originally desired.

On top of that, the small trait `S3Location` exists for extensions that want to: 
 - have the basic fields directly accessible (protocol, bucketName, path)
 - want to have access to the `simpleS3LocationString` - this string is the default implementation for the trait and even if other extensions may include some other data (region, ...), that should not be reflected in this simpleS3Location string representation.
 
I hope this is now acceptable.